File: niko-studio/src/knowledge/services/embedding_service.py
"""
Embedding 服务实现

统一封装多 Provider，支持缓存、批量处理和相似度计算。
"""

import asyncio
import math

from .models import (
    EmbeddingRequest,
    EmbeddingResponse,
    ModelTier,
    ProviderType,
    TokenUsage,
    EmbeddingError,
)
from .protocols import EmbeddingProvider, EmbeddingCache


class EmbeddingServiceImpl:
    """Embedding 服务实现

    提供统一的 Embedding 调用接口，支持:
    - 多 Provider 管理
    - Embedding 缓存
    - 批量处理
    - 相似度计算
    """

    def __init__(
        self,
        providers: dict[ProviderType, EmbeddingProvider],
        default_provider: ProviderType = ProviderType.OPENAI,
        cache: EmbeddingCache | None = None,
        default_model: str | None = None,
    ):
        """初始化 Embedding 服务

        Args:
            providers: Provider 实例映射
            default_provider: 默认 Provider 类型
            cache: 可选的缓存实现
            default_model: 默认模型名称
        """
        self._providers = providers
        self._default_provider = default_provider
        self._cache = cache
        self._default_model = default_model

    def _get_provider(self, provider_type: ProviderType | None = None) -> EmbeddingProvider:
        """获取 Provider 实例"""
        ptype = provider_type or self._default_provider
        if ptype not in self._providers:
            available = list(self._providers.keys())
            if not available:
                raise EmbeddingError("No providers available")
            ptype = available[0]
        return self._providers[ptype]

    def _get_default_model(self, provider: EmbeddingProvider) -> str:
        """获取默认模型"""
        if self._default_model:
            return self._default_model
        # 根据 Provider 类型返回默认模型
        if provider.provider_type == ProviderType.OPENAI:
            return "text-embedding-3-small"
        if provider.provider_type == ProviderType.LOCAL:
            return "BAAI/bge-small-zh-v1.5"
        return "text-embedding-3-small"

    async def embed(
        self,
        text: str,
        *,
        model: str | None = None,
        provider: ProviderType | None = None,
    ) -> list[float]:
        """生成单个文本的向量表示

        Args:
            text: 待编码的文本
            model: 模型名称
            provider: 指定 Provider

        Returns:
            向量表示
        """
        embeddings = await self.embed_batch([text], model=model, provider=provider)
        return embeddings[0]

    async def embed_batch(
        self,
        texts: list[str],
        *,
        model: str | None = None,
        batch_size: int = 100,
        provider: ProviderType | None = None,
    ) -> list[list[float]]:
        """批量生成文本的向量表示

        Args:
            texts: 待编码的文本列表
            model: 模型名称
            batch_size: 每批次处理的文本数量
            provider: 指定 Provider

        Returns:
            向量表示列表
        """
        if not texts:
            return []

        emb_provider = self._get_provider(provider)
        model_name = model or self._get_default_model(emb_provider)

        # 检查缓存
        cache_results: dict[str, list[float] | None] = {}
        texts_to_embed: list[str] = []
        text_indices: dict[str, int] = {}

        if self._cache:
            cache_results = await self._cache.get_batch(texts, model_name)
            for i, text in enumerate(texts):
                text_indices[text] = i
                if cache_results.get(text) is None:
                    texts_to_embed.append(text)
        else:
            texts_to_embed = texts
            for i, text in enumerate(texts):
                text_indices[text] = i

        # 分批处理未缓存的文本
        new_embeddings: dict[str, list[float]] = {}
        for i in range(0, len(texts_to_embed), batch_size):
            batch = texts_to_embed[i:i + batch_size]
            response = await emb_provider.embed(batch, model_name)
            for text, embedding in zip(batch, response.embeddings):
                new_embeddings[text] = embedding

        # 更新缓存
        if self._cache and new_embeddings:
            await self._cache.set_batch(new_embeddings, model_name)

        # 合并结果，保持原始顺序
        results: list[list[float]] = [[] for _ in texts]
        for text in texts:
            idx = text_indices[text]
            if text in new_embeddings:
                results[idx] = new_embeddings[text]
            elif cache_results.get(text):
                results[idx] = cache_results[text]

        return results

    async def embed_with_metadata(self, request: EmbeddingRequest) -> EmbeddingResponse:
        """生成向量表示，包含完整元数据"""
        emb_provider = self._get_provider()
        model_name = request.model_override or self._get_default_model(emb_provider)

        # 检查缓存
        cache_hits = 0
        texts_to_embed = request.texts
        cached_embeddings: dict[str, list[float]] = {}

        if self._cache:
            cache_results = await self._cache.get_batch(request.texts, model_name)
            texts_to_embed = []
            for text in request.texts:
                if cache_results.get(text):
                    cached_embeddings[text] = cache_results[text]
                    cache_hits += 1
                else:
                    texts_to_embed.append(text)

        # 调用 Provider
        if texts_to_embed:
            response = await emb_provider.embed(
                texts_to_embed,
                model_name,
                dimensions=request.dimensions,
            )

            # 更新缓存
            if self._cache:
                new_items = dict(zip(texts_to_embed, response.embeddings))
                await self._cache.set_batch(new_items, model_name)
                cached_embeddings.update(new_items)

            # 合并结果
            embeddings = [cached_embeddings.get(t, []) for t in request.texts]

            return EmbeddingResponse(
                embeddings=embeddings,
                model_used=model_name,
                provider=emb_provider.provider_type,
                dimensions=response.dimensions,
                usage=response.usage,
                latency_ms=response.latency_ms,
                cache_hits=cache_hits,
            )
        else:
            # 全部命中缓存
            embeddings = [cached_embeddings[t] for t in request.texts]
            dimensions = len(embeddings[0]) if embeddings else 0

            return EmbeddingResponse(
                embeddings=embeddings,
                model_used=model_name,
                provider=emb_provider.provider_type,
                dimensions=dimensions,
                usage=TokenUsage(),
                latency_ms=0,
                cache_hits=cache_hits,
            )

    def similarity(
        self,
        embedding1: list[float],
        embedding2: list[float],
    ) -> float:
        """计算两个向量的余弦相似度

        Args:
            embedding1: 第一个向量
            embedding2: 第二个向量

        Returns:
            余弦相似度 (-1 到 1)
        """
        if len(embedding1) != len(embedding2):
            raise ValueError("Embeddings must have the same dimensions")

        dot_product = sum(a * b for a, b in zip(embedding1, embedding2))
        norm1 = math.sqrt(sum(a * a for a in embedding1))
        norm2 = math.sqrt(sum(b * b for b in embedding2))

        if norm1 == 0 or norm2 == 0:
            return 0.0

        return dot_product / (norm1 * norm2)

    def get_dimensions(self, model: str | None = None) -> int:
        """获取模型的向量维度"""
        provider = self._get_provider()
        model_name = model or self._get_default_model(provider)
        return provider.get_dimensions(model_name)

File: src-ts/knowledge/embedding-service.ts
/**
 * Knowledge module - embedding service
 *
 * Unified multi-provider embedding service with caching, batch processing,
 * and similarity calculation.
 */

import type { EmbeddingCache } from '../protocols/embedding';
import type {
  EmbeddingRequest,
  EmbeddingResponse,
  EmbeddingProvider,
  TokenUsage,
} from './models';
import {
  ProviderType,
  createTokenUsage,
  createEmbeddingResponse,
  EmbeddingError,
} from './models';

/**
 * Embedding service implementation
 *
 * Provides a unified embedding API supporting:
 * - Multi-provider management
 * - Embedding cache
 * - Batch processing
 * - Similarity calculation
 */
export class EmbeddingServiceImpl {
  private readonly _providers: Map<string, EmbeddingProvider>;
  private readonly _defaultProvider: string;
  private readonly _cache: EmbeddingCache | null;
  private readonly _defaultModel: string | null;

  constructor(params: {
    providers: Map<string, EmbeddingProvider>;
    defaultProvider?: string;
    cache?: EmbeddingCache | null;
    defaultModel?: string | null;
  }) {
    this._providers = params.providers;
    this._defaultProvider = params.defaultProvider ?? ProviderType.LOCAL;
    this._cache = params.cache ?? null;
    this._defaultModel = params.defaultModel ?? null;
  }

  /**
   * Get provider instance by type
   */
  private _getProvider(providerType?: string | null): EmbeddingProvider {
    let ptype = providerType ?? this._defaultProvider;
    const provider = this._providers.get(ptype);
    if (!provider) {
      const available = Array.from(this._providers.keys());
      if (available.length === 0) {
        throw new EmbeddingError('No providers available');
      }
      ptype = available[0];
      return this._providers.get(ptype)!;
    }
    return provider;
  }

  /**
   * Get default model name for a provider
   */
  private _getDefaultModel(provider: EmbeddingProvider): string {
    if (this._defaultModel) return this._defaultModel;
    if (provider.providerType === ProviderType.OPENAI) {
      return 'text-embedding-3-small';
    }
    if (provider.providerType === ProviderType.LOCAL) {
      return 'BAAI/bge-small-en-v1.5';
    }
    return 'text-embedding-3-small';
  }

  /**
   * Generate vector representation for a single text
   */
  async embed(
    text: string,
    options?: { model?: string | null; provider?: string | null },
  ): Promise<number[]> {
    const embeddings = await this.embedBatch([text], {
      model: options?.model,
      provider: options?.provider,
    });
    return embeddings[0];
  }

  /**
   * Batch generate vector representations for texts
   */
  async embedBatch(
    texts: string[],
    options?: {
      model?: string | null;
      batchSize?: number;
      provider?: string | null;
    },
  ): Promise<number[][]> {
    if (texts.length === 0) return [];

    const embProvider = this._getProvider(options?.provider);
    const modelName = options?.model ?? this._getDefaultModel(embProvider);
    const batchSize = options?.batchSize ?? 100;

    // Check cache
    let cacheResults: Record<string, number[] | null> = {};
    const textsToEmbed: string[] = [];
    const textIndices: Map<string, number> = new Map();

    if (this._cache) {
      cacheResults = await this._cache.getBatch(texts, modelName);
      for (let i = 0; i < texts.length; i++) {
        textIndices.set(texts[i], i);
        if (cacheResults[texts[i]] == null) {
          textsToEmbed.push(texts[i]);
        }
      }
    } else {
      for (let i = 0; i < texts.length; i++) {
        textIndices.set(texts[i], i);
      }
      textsToEmbed.push(...texts);
    }

    // Process uncached texts in batches
    const newEmbeddings: Map<string, number[]> = new Map();
    for (let i = 0; i < textsToEmbed.length; i += batchSize) {
      const batch = textsToEmbed.slice(i, i + batchSize);
      const embeddings = await embProvider.embed(batch, modelName);
      for (let j = 0; j < batch.length; j++) {
        newEmbeddings.set(batch[j], embeddings[j]);
      }
    }

    // Update cache
    if (this._cache && newEmbeddings.size > 0) {
      const items: Record<string, number[]> = {};
      for (const [k, v] of newEmbeddings) {
        items[k] = v;
      }
      await this._cache.setBatch(items, modelName);
    }

    // Merge results, preserving original order
    const results: number[][] = new Array(texts.length).fill(null as any) as number[][];
    for (const text of texts) {
      const idx = textIndices.get(text)!;
      if (newEmbeddings.has(text)) {
        results[idx] = newEmbeddings.get(text)!;
      } else if (cacheResults[text]) {
        results[idx] = cacheResults[text]!;
      }
    }

    return results;
  }

  /**
   * Generate vector representations with full metadata
   */
  async embedWithMetadata(request: EmbeddingRequest): Promise<EmbeddingResponse> {
    const embProvider = this._getProvider();
    const modelName = request.modelOverride ?? this._getDefaultModel(embProvider);

    // Check cache
    let cacheHits = 0;
    let textsToEmbed = [...request.texts];
    const cachedEmbeddings: Map<string, number[]> = new Map();

    if (this._cache) {
      const cacheResults = await this._cache.getBatch(request.texts, modelName);
      textsToEmbed = [];
      for (const text of request.texts) {
        if (cacheResults[text]) {
          cachedEmbeddings.set(text, cacheResults[text]!);
          cacheHits++;
        } else {
          textsToEmbed.push(text);
        }
      }
    }

    // Call provider
    if (textsToEmbed.length > 0) {
      const embeddings = await embProvider.embed(textsToEmbed, modelName);

      // Update cache
      if (this._cache) {
        const newItems: Record<string, number[]> = {};
        for (let i = 0; i < textsToEmbed.length; i++) {
          newItems[textsToEmbed[i]] = embeddings[i];
          cachedEmbeddings.set(textsToEmbed[i], embeddings[i]);
        }
        await this._cache.setBatch(newItems, modelName);
      }

      // Merge results
      const allEmbeddings = request.texts.map(t => cachedEmbeddings.get(t) ?? []);
      const actualDimensions = allEmbeddings.length > 0 && allEmbeddings[0] ? allEmbeddings[0].length : 0;

      return createEmbeddingResponse({
        embeddings: allEmbeddings,
        modelUsed: modelName,
        provider: embProvider.providerType as any,
        dimensions: actualDimensions,
        usage: createTokenUsage(),
        latencyMs: 0,
        cacheHits,
      });
    } else {
      // All cache hits
      const embeddings = request.texts.map(t => cachedEmbeddings.get(t) ?? []);
      const dimensions = embeddings.length > 0 && embeddings[0] ? embeddings[0].length : 0;

      return createEmbeddingResponse({
        embeddings,
        modelUsed: modelName,
        provider: embProvider.providerType as any,
        dimensions,
        usage: createTokenUsage(),
        latencyMs: 0,
        cacheHits,
      });
    }
  }

  /**
   * Calculate cosine similarity between two vectors
   *
   * @returns Cosine similarity (-1 to 1)
   */
  similarity(embedding1: number[], embedding2: number[]): number {
    if (embedding1.length !== embedding2.length) {
      throw new Error('Embeddings must have the same dimensions');
    }

    let dotProduct = 0;
    let norm1 = 0;
    let norm2 = 0;

    for (let i = 0; i < embedding1.length; i++) {
      dotProduct += embedding1[i] * embedding2[i];
      norm1 += embedding1[i] * embedding1[i];
      norm2 += embedding2[i] * embedding2[i];
    }

    norm1 = Math.sqrt(norm1);
    norm2 = Math.sqrt(norm2);

    if (norm1 === 0 || norm2 === 0) return 0.0;

    return dotProduct / (norm1 * norm2);
  }

  /**
   * Get vector dimensions for a model
   */
  async getDimensions(model?: string | null): Promise<number> {
    const provider = this._getProvider();
    const modelName = model ?? this._getDefaultModel(provider);
    // Embed a test string to determine dimensions
    const [embedding] = await provider.embed(['test'], modelName);
    return embedding.length;
  }
}
